### PR TITLE
int-tests: Add mock websocket server, client, and a basic testcase

### PR DESCRIPTION
**Pull request by suksw**
### Purpose
- Add a basic mock WebSocket echo server (supports ping, pong)
- Create a WebSocket client to send a list of strings
- Add a basic websocket test only to invoke a WebSocket endpoint with a prod oauth2 token
- Update test util methods to support updating asyncAPI definitions
- Remove test dependencies that are currently not in use
- Update logging in the mockbackend to use slf4j (mockbackend logs started to appear after using LoggerFactory instead of LogManager)

### Issues

Fixes https://github.com/wso2/product-microgateway/issues/2675

### References
- https://github.com/netty/netty/blob/4.1/example/src/main/java/io/netty/example/http/websocketx/server/WebSocketServer.java
- https://github.com/netty/netty/blob/2b811c7895e8c11d64b25fa6c1308846392e2be0/example/src/main/java/io/netty/example/http/websocketx/server/WebSocketServerInitializer.java
- https://github.com/netty/netty/blob/4.1/example/src/main/java/io/netty/example/http/websocketx/client/WebSocketClient.java
- https://github.com/netty/netty/blob/4.1/example/src/main/java/io/netty/example/http/websocketx/client/WebSocketClientHandler.java
- https://medium.com/@irunika/how-to-write-a-http-websocket-server-using-netty-f3c136adcba9
- https://github.com/wso2/product-microgateway/blob/f8e364fcbfe188ae10c8ae99e204c364c492a5c2/integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/util/HttpClientRequest.java#L296

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/apim/utils/PublisherUtils.java [hunk 1]
```diff
@@ -115,30 +115,54 @@ public static String createAPIUsingOAS(JSONObject apiProperties,
...
+            throw new CCTestException("Error while updating OpenAPI", e);
         } catch (IOException e) {
             log.error("Error occurred while reading OpenAPI definition for: {}", openAPIFileName);
             throw new CCTestException("Error while reading OpenAPI definition", e);
         }
-        return responseApiId;
+    }
+
+    /**
+     * Updates the OpenAPI definition of an already created AsyncAPI
```
Comment (suksw) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/apim/utils/PublisherUtils.java:142]:
```suggestion
     * Updates the AsyncAPI definition of an already created AsyncAPI
```

File: integration/mock-backend-server/src/main/java/org/wso2/choreo/connect/mockbackend/async/MockAsyncServer.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+            ServerBootstrap b = new ServerBootstrap();
+            b.group(bossGroup, workerGroup)
+                    .channel(NioServerSocketChannel.class)
+                    .handler(new LoggingHandler(LogLevel.INFO))
+                    .childHandler(new WebSocketServerInitializer(null));
+
+            Channel ch = b.bind(serverPort).sync().channel();
+            ch.closeFuture().sync();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
```
Comment (suksw) [inline integration/mock-backend-server/src/main/java/org/wso2/choreo/connect/mockbackend/async/MockAsyncServer.java:63]: replace with a log

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/util/websocket/WsClient.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+            log.info("Websocket client initiating connection");
+            Channel ch = b.connect(uri.getHost(), uri.getPort()).sync().channel();
+            Utils.delay(1000, "aa");
+            handler.handshakeFuture().sync();
+
+            log.info("Websocket client handshake complete");
+
+            sendMessages(ch, messages);
+        } catch (URISyntaxException e) {
+            e.printStackTrace();
```
Comment (suksw) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/util/websocket/WsClient.java:120]: convert to logs

File: integration/test-integration/src/test/resources/asyncAPIs/websocket_basic_asyncAPI.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    publish: {}
+components:
+  securitySchemes:
+    oauth2:
+      type: oauth2
+      flows:
+        implicit:
+          authorizationUrl: 'http://localhost:9999'
+          scopes: {}
+          x-scopes-bindings: {}
```
Comment (suksw) [inline integration/test-integration/src/test/resources/asyncAPIs/websocket_basic_asyncAPI.yaml:20]: Add new line

File: integration/mock-backend-server/pom.xml [hunk 5]
```diff
@@ -42,6 +42,10 @@
             <groupId>org.apache.logging.log4j</groupId>
             <artifactId>log4j-core</artifactId>
         </dependency>
+        <dependency>
+            <groupId>org.slf4j</groupId>
+            <artifactId>slf4j-simple</artifactId>
+        </dependency>
```
Comment (VajiraPrabuddhaka) [inline integration/mock-backend-server/pom.xml:48]: Aren't we use log4j for logging in integration tests?

Comment (suksw) [reply integration/mock-backend-server/pom.xml:48]: For some reason the logs for mockbackend does not get printed when I used log4j. Then when changed to a slf4j logger created using LogFactory the logs appear for the container. I just double checked by using a log4j logger again. This is still the case.

File: integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/withapim/websocket/WsBasicTestCase.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    @BeforeClass(alwaysRun = true, description = "Create access token and define endpoint URL")
+    void setEnvironment() throws Exception {
+        super.initWithSuperTenant();
+
+        // Get App ID and API ID
+        String applicationId = ApimResourceProcessor.applicationNameToId.get(APPLICATION_NAME);
+
+        String accessToken = StoreUtils.generateUserAccessToken(apimServiceURLHttps, applicationId,
+                user, storeRestClient);
+        requestHeaders.put(TestConstant.AUTHORIZATION_HEADER, "Bearer " + accessToken);
```
Comment (AmaliMatharaarachchi) [inline integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/withapim/websocket/WsBasicTestCase.java:54]: shall we add a test for invalid/no auth token as well. also a test when x-wso2-disable-security is provided for the ws api.

Comment (suksw) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/withapim/websocket/WsBasicTestCase.java:54]: +1

Comment (suksw) [reply integration/test-integration/src/test/java/org/wso2/choreo/connect/tests/testcases/withapim/websocket/WsBasicTestCase.java:54]: looks like I haven't decoded the resource level disable security extensions for asyncAPIs. Created an issue here. https://github.com/wso2/product-microgateway/issues/2708 Will send the test together with the fix

**Review by AmaliMatharaarachchi — APPROVED** [review]

**Pull request merged**